### PR TITLE
Move release from travis to github actions

### PR DESCRIPTION
**Pull request by ssbarnea**
- create packaging environment that produces deliverables
- removes travis integration
- configures github actions to call packaging after all the other
  tests are passing and to make a release if tag is present
- test.pypi.org is kept commented until we sort credentials for it
- no credentials are needed as they are already configured in github
- enables a daily cron run of CI, so we know if something goes wrong for external reasons

### CODE REVIEW

**Comment by ssbarnea** [comment]
@BeyondEvil @gnikonorov That is ready to ship!

**Comment by BeyondEvil** [comment]
I've started the review. But since the GHA is all new to me, it's going to take a while.

File: .github/workflows/actions.yml
Comment (gnikonorov) [inline .github/workflows/actions.yml:15]: You're just introducing a daily CI build here, right?

File: .github/workflows/actions.yml
Comment (gnikonorov) [inline .github/workflows/actions.yml:166]: just out of curiosity, why do we want to drop all tags for non tag create events?

**Review by gnikonorov — COMMENTED** [review]
I did a quick skim now. Overall I think it looks good, I just had one question out of curiosity. I have to step away, but I'll do a proper review when I'm back.

File: .github/workflows/actions.yml
Comment (nicoddemus) [inline .github/workflows/actions.yml:158]: Also not clear to me why we are deleting tags here, but I think just setting this to `0` is OK: pytest-html is a tiny repo so cloning the entire history is fine IMHO.

**Review by nicoddemus — COMMENTED** [review]
Agree, overall looks good, I have the same questions regarding deleting tags though.

**Review by gnikonorov — CHANGES_REQUESTED** [review]
Since this will officially deprecate our usage of travis, can you remove any references to it as part of this commit or open an issue after this merges so we don't forget? I see the following references to it in the codebase:

```
$ git grep travis
README.rst:.. image:: https://img.shields.io/travis/pytest-dev/pytest-html.svg
README.rst:   :target: https://travis-ci.org/pytest-dev/pytest-html/
development.rst:All pull requests and merges are tested in `Travis CI <https://travis-ci.org/>`_
development.rst:based on the ``.travis.yml`` file.
development.rst:Usually, a link to your specific travis build appears in pull requests, but if
development.rst:`pull requests page <https://travis-ci.org/pytest-dev/pytest-html/pull_requests>`_
development.rst:8. Done. You can monitor the progress on `Travis <https://travis-ci.org/pytest-dev/pytest-html/>`_
$
```

File: .github/workflows/actions.yml
Comment (ssbarnea) [reply .github/workflows/actions.yml:15]: Exactly.

File: .github/workflows/actions.yml
Comment (ssbarnea) [reply .github/workflows/actions.yml:166]: I do not remember exactly as I was not the original author of this task, maybe @webknjaz could help here?

My guess is to avoid getting setuptools-scm confused about which tag it would use for producing the version, especially as it is possible to add a tag somewhere back-in-time (before more recent tag in the branch).

**Comment by ssbarnea** [comment]
> Since this will officially deprecate our usage of travis, can you remove any references to it as part of this commit or open an issue after this merges so we don't forget? I see the following references to it in the codebase:
> 
> ```
> $ git grep travis
> README.rst:.. image:: https://img.shields.io/travis/pytest-dev/pytest-html.svg
> README.rst:   :target: https://travis-ci.org/pytest-dev/pytest-html/
> development.rst:All pull requests and merges are tested in `Travis CI <https://travis-ci.org/>`_
> development.rst:based on the ``.travis.yml`` file.
> development.rst:Usually, a link to your specific travis build appears in pull requests, but if
> development.rst:`pull requests page <https://travis-ci.org/pytest-dev/pytest-html/pull_requests>`_
> development.rst:8. Done. You can monitor the progress on `Travis <https://travis-ci.org/pytest-dev/pytest-html/>`_
> $
> ```

Indeed, I forgot the cleanup travis references. I will update the PR.

File: .github/workflows/actions.yml
Comment (webknjaz) [inline .github/workflows/actions.yml:180]: How about
```suggestion
      #       false &&
      #       github.event_name == 'push' &&
```
instead of commenting out the whole thing?

File: .github/workflows/actions.yml
Comment (webknjaz) [reply .github/workflows/actions.yml:166]: @ssbarnea yes, that would be one subcase.
Although, I've been doing this in particular to avoid a slightly different race condition.
When the tags are pushed, this generates 2 workflows: for tag `push` and for tag `create`.
This means that whenever both of these workflows will run at about the same time and the checkout action will get the tag and the commits. And then, setuptools-scm will generate a version with that tag exactly. And this is fine because later in the workflow, the publish step is executed under only one of these cases so both jobs won't attempt to upload. That's why this step is skipped for these cases.
Now, there's a third job that may run separately in the context of the same commit: `push` to master. And if that commit is also tagged (for example, if the tagging happens fast after pushing the same commit to master, before some of the jobs did checkout), setuptools-scm will generate a version based on that tag. This is bad when that "Publish to test.pypi.org" step is uncommented because it'll result in two jobs attempting to publish an artifact with the same name to TestPyPI and one of them will fail (with no guarantee which one). And so for this case, I'm removing any tags that may be present in the HEAD which eliminates the race condition.

File: tox.ini
Comment (webknjaz) [inline tox.ini:46]: Isn't this used for `setup.py check` that is not executed here?

File: tox.ini
Comment (webknjaz) [inline tox.ini:49]: This is a transitive dependency of `pep517`, it probably doesn't matter what version it is.

**Comment by webknjaz** [comment]
FTR these days I'd rely on https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/ instead of `create` because `create` requires post-checks in runtime and also runs on branch creation too.

File: .github/workflows/actions.yml
Comment (gnikonorov) [reply .github/workflows/actions.yml:15]: Thank you for clarifying!

File: .github/workflows/actions.yml
Comment (gnikonorov) [reply .github/workflows/actions.yml:166]: Thank you for clarifying @webknjaz. Since @nicoddemus also had the same question, I'll ask him to resolve this if this answer works for him

File: development.rst
Comment (gnikonorov) [inline development.rst:27]: nitpick: 'inside the ``.github`` folder.'

File: development.rst
Comment (gnikonorov) [inline development.rst:30]: nitpick: 'use the dropdown option, close and reopen the pull-request or just update the branch'

**Review by gnikonorov — CHANGES_REQUESTED** [review]
Thank you for cleaning up the docs! I have a few nitpicks, but don't mind approving after @webknjaz's feedback is addressed.

**Comment by webknjaz** [comment]
FWIW none of my feedback is blocking, it's rather informative.

**Comment by gnikonorov** [comment]
> FWIW none of my feedback is blocking, it's rather informative.

Thank you for clarifying @webknjaz! 

@ssbarnea let me know then when you've addressed all the feedback you wanted to and I'll approve. I'd like to wait for @BeyondEvil to also approve before we merge

File: .github/workflows/actions.yml
Comment (nicoddemus) [reply .github/workflows/actions.yml:166]: Thanks @webknjaz for the explanation.

Under normal circumstances in our release workflow the race condition cannot really happen, because the tag will be created in the last commit of the release PR, and only after that is green (and hence the package is published), the commit will be merged to `master`, not as fast-forward but as a normal merge commit.

So I'm under the opinion that this is a bit of complication that really doesn't bring anything to the table for `pytest-dev/pytest-html`, but it is also harmless so feel free to go ahead with it. 👍

**Comment by nicoddemus** [comment]
> FTR these days I'd rely on github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch instead of create because create requires post-checks in runtime and also runs on branch creation too.

I also think this would be nice, but just as @webknjaz, this not blocking for me. 😁

**Review by nicoddemus — APPROVED** [review]

**Comment by ssbarnea** [comment]
I managed to setup testpypi and I will update this PR tomorrow.

**Comment by gnikonorov** [comment]
Hey @ssbarnea can we merge this in first, and then add `testpypi` in a follow up PR?

File: tox.ini
Comment (ssbarnea) [reply tox.ini:46]: Added

**Comment by ssbarnea** [comment]
I had to rebase, fixed additional commend. To avoid delaying this even more lets merge it as soon we get the review and we will do other improvements in follow-ups.

**Comment by gnikonorov** [comment]
> I had to rebase, fixed additional commend. To avoid delaying this even more lets merge it as soon we get the review and we will do other improvements in follow-ups.

Of course @ssbarnea! I'll take a look today. Thank you

**Review by gnikonorov — APPROVED** [review]
LGTM!

As @ssbarnea mentioned we can get any improvements we need in subsequent PRs. I think this is fine as is and will let us start releasing with the new workflow.

I would like one more person to approve before merge, however. @BeyondEvil  can you take a look as well?

File: .github/workflows/actions.yml
Comment (BeyondEvil) [inline .github/workflows/actions.yml:147]: Is there a reason why we're using 3.6 and not a newer version @ssbarnea @gnikonorov ?

Comment (webknjaz) [reply .github/workflows/actions.yml:147]: I'm pretty sure Sorin wanted to support the oldest possible version that the devs/contributors may use.

Comment (ssbarnea) [reply .github/workflows/actions.yml:147]: Yes, use of oldest supported is usually better than newest in order to avoid regressions. Based on experience using oldest proved to catch more issues than the newest, especially in this area. 

Any CI failure that prevents a bad release is better than making a user unhappy. Note to myself: add comments on action tasks like this documenting their purpose.

Comment (BeyondEvil) [reply .github/workflows/actions.yml:147]: Thanks @ssbarnea! Makes sense. 👍

**Pull request merged**
